Form field icons load in default 'grey' state

ErisDS opened the pull request:
I'm totes a :monkey_face: at a :musical_keyboard: with this stuff. The 'proper' fix is likely collapsing this state along with errors into a single concept of validation state, but this resolves the issue such that icons load grey, and don't change colour until some sort of validation has happened.

ref #5653

- validations can be in default, success or error state
- adds check for 'hasValidated' if the validations haven't fired yet, the field is in the default state

kevinansfield commented:
Looks good for a temporary fix :+1: I'll pull down and give it a proper workout in the morning - I want to check the behaviour where properties are validated individually but the validated object is marked `hasValidated` as a whole.

kevinansfield commented:
The only place I've found an issue is on setup/two - as soon as one field is marked as invalid all of the other fields turn green. Perhaps this could be worked around when tackling the "Do not validate empty form fields on setup" issue?

<img width="423" alt="screen shot 2015-08-28 at 09 19 51" src="https://cloud.githubusercontent.com/assets/415/9542144/1e339362-4d66-11e5-983a-42c08dbd1a9c.png">

ErisDS commented:
Ah good catch cc @cobbspur - let's see how that issue interacts with this one? If it's still buggy I think I can fix it.

ErisDS commented:
Rebased against master & all of the recent changes that have been made to these forms. 

I have also made a small change, such that `hasValidated` is an array, rather than a boolean. It now stores & checks against the names of the properties which have been validated, meaning the setup/two screen behaves properly.

This seems to be working nicely with David's changes to prevent validations firing for empty fields on focus out, and I think it feels much nicer now.

ErisDS commented:
Have updated this again, as it was a little too simplistic to cope properly with all the cases, including the way the signin form is validated as a whole rather than as fields, and some things were broken with the settings pages: general, users & team.

It is now passing tests, and I've given it a much more thorough look over.

kevinansfield commented on core/client/app/controllers/setup/three.js:
```diff
@@ -6,8 +6,9 @@ export default Ember.Controller.extend({
+    hasValidated: [],
     users: '',
-    ownerEmail: Ember.computed.alias('two.email'),
+    ownerEmail: 'hannah@ghost.org',
```
Guessing this has been left in by accident?

ErisDS replied:
Whoops

kevinansfield commented:
I think there's a regression here somewhere, on `setup/two` the form-groups no longer have the `error` class applied when trying to submit with invalid data.

<img width="418" alt="screen shot 2015-08-31 at 09 41 00" src="https://cloud.githubusercontent.com/assets/415/9575083/a94cbbd0-4fc4-11e5-8718-74259eee9ea9.png">

ErisDS commented:
Updated to resolve that (also happening on signup). This unquestionably needs revisiting now that we've got a clear idea of all the weird edges that need to be accounted for.

cobbspur commented:
A few points

signin page:
- You can submit with an email address of 'd' and it validates succesfully? 
- If you try to signin as 'd' (or any none email address in previous item, and no password you get the  success icon if you then add a valid email address and click forgot? - you seem to get a darker green icon http://recordit.co/Vix4zr1kP5

Team page:
- if you invite a user with no email address or 'dddd' or 'dddd@dddd' you get no indication that something was wrong

Setup - step 2
- Add a valid email address and focus off email input - no validation takes place. Then change it to an invalid email address and you correctly get the error. Now change back to valid email address and validation is successful.
- Same as above for password
- Using autocomplete on email address results in ![](http://puu.sh/jVhP6/6d13855b84.png)
- Validation does not occur for full name or blog title until email or password successfully validate

ErisDS commented:
Updated this again. 

- The reported bugs with signin are in master, so fixed separately.
- The issues with setup/two were to do with me trying to watch a bog-standard array (not a thing) so I've switched `hasValidated` to be an Ember.Array, and am using all the proper functions.

Team page I knew about - this never got updated to use inline validations, and we've removed the error notifications used instead. Going to try to look at this now.

cobbspur commented:
tested the setup flow -everything seems to working nicely now :+1:
